src/planning/recipe-index.ts

```diff
@@ -308,10 +315,16 @@ export class RecipeIndex {
    * - `matchingHandleConn` - a handle connection of a particle, whose slot connection is explored
    *    as a potential match to a slot above.
    */
-  private _fatesAndDirectionsMatch(slotHandleConn, matchingHandleConn): boolean {
+  private _fatesAndDirectionsMatch(slotHandleConn: HandleConnection, matchingHandleConn: HandleConnection): boolean {
     const matchingHandle = matchingHandleConn.handle;
     const allMatchingHandleConns = matchingHandle ? matchingHandle.connections : [matchingHandleConn];
-    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => ['out', 'inout'].includes(conn.direction));
+    const matchingDirection: Direction[] = ['out', 'inout'];
+    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => conn.direction && matchingDirection.includes(conn.direction));
+
+    if (!slotHandleConn.handle) {
```

Comment:
can this be an assert instead?

Reply:
Typescript will narrow a nullable/undefined type to non-nullable/defined when you use an if statement or other control flow.  Sadly assert() does not provide that.

The `preconditions.ts` class has checkDefined() and checkNotNull() and they will narrow nullness.  So far have only used those for testing.  WDYT?  Move those up a level and make them available for general use?

Reply:
I think it'll be nice to use them in the code as well - it's easy to mask bugs with the if statements, and having both - if and assert is too verbose.